components/cam/src/physics/cam/phys_grid.F90

```diff
@@ -4540,13 +4704,81 @@ subroutine create_chunks(opt, chunks_per_thread)
 !
 !  Allocate other work space
 !
-   allocate( nvsmpthreads(0:nvsmp-1) )
-   allocate( nvsmpchunks (0:nvsmp-1) )
-   allocate( maxcol_chk  (0:nvsmp-1) )
-   allocate( maxcol_chks (0:nvsmp-1) )
-   allocate( cid_offset  (0:nvsmp-1) )
-   allocate( local_cid   (0:nvsmp-1) )
-   allocate( cols_d    (1:maxblksiz) )
+   allocate( nvsmpthreads  (0:nvsmp-1) )
+   allocate( nvsmpchunks   (0:nvsmp-1) )
+   allocate( pcols_vsmp    (0:nvsmp-1) )
+   allocate( maxcol_chk    (0:nvsmp-1) )
+   allocate( maxcol_chks   (0:nvsmp-1) )
+   allocate( maxblksiz_vsmp(0:nvsmp-1) )
+   allocate( cid_offset    (0:nvsmp-1) )
+   allocate( local_cid     (0:nvsmp-1) )
+   allocate( cols_d      (1:maxblksiz) )
+
+!
+! Calculate number of threads available in each virtual SMP node. 
+!
+   nvsmpthreads(:) = 0
+   do p=0,npes-1
+      smp = proc_vsmp_map(p)
+      nvsmpthreads(smp) = nvsmpthreads(smp) + npthreads(p)
+   enddo
+
+!
+! Calculate maximum block size for each virtual SMP
+!
+   maxblksiz_vsmp(:) = 0
+   do j=firstblock,lastblock
+      p = get_block_owner_d(j)
+      smp = proc_vsmp_map(p)
+      blksiz = get_block_gcol_cnt_d(j)
+      if (blksiz > maxblksiz_vsmp(smp)) maxblksiz_vsmp(smp) = blksiz
+   enddo
+
+!
+! Calculate pcols for chunks in each virtual SMP
```

Comment:
I also agree with @ambrad's sentiment that this block should ideally be it's own routine that could be covered with a  unit test. Same goes for the section below that utilizes the optional column costs.

Reply:
Okay for this to be future work?

Reply:
yea sure, just wanted to vocalize my support for more modularity.